File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/base/BaseDialogSelection.java
package com.github.RooneyAndShadows.lightbulb.dialogs.base;

import java.util.ArrayList;
import java.util.HashMap;

public abstract class BaseDialogSelection<SelectionType> {
    private final static String SELECTION_CURRENT = "SELECTION_CURRENT";
    private final static String SELECTION_DRAFT = "SELECTION_DRAFT";
    private final HashMap<String, SelectionType> selection = new HashMap<>();
    private final ArrayList<PickerSelectionListeners<SelectionType>> selectionListeners = new ArrayList<>();

    public BaseDialogSelection(SelectionType current, SelectionType draft) {
        selection.put(SELECTION_CURRENT, current);
        selection.put(SELECTION_DRAFT, draft);
    }

    public abstract boolean compareValues(SelectionType v1, SelectionType v2);

    public abstract boolean hasCurrentSelection();

    public abstract boolean hasDraftSelection();

    public SelectionType getCurrentSelection() {
        return selection.get(SELECTION_CURRENT);
    }

    public SelectionType getDraftSelection() {
        return selection.get(SELECTION_DRAFT);
    }

    public void addSelectionListeners(PickerSelectionListeners<SelectionType> selectionListeners) {
        this.selectionListeners.add(selectionListeners);
    }

    public void setCurrentSelection(SelectionType newValue) {
        setCurrentSelection(newValue, true);
    }

    public void setDraftSelection(SelectionType newValue) {
        setDraftSelection(newValue, true);
    }

    public void setCurrentSelection(SelectionType newValue, boolean notify) {
        SelectionType current = selection.get(SELECTION_CURRENT);
        if (compareValues(current, newValue))
            return;
        selection.remove(SELECTION_CURRENT);
        selection.put(SELECTION_CURRENT, newValue);
        if (notify)
            for (PickerSelectionListeners<SelectionType> listener : selectionListeners)
                listener.onCurrentSelectionChangedListener(newValue, current);
    }

    public void setDraftSelection(SelectionType newValue, boolean notify) {
        SelectionType draft = selection.get(SELECTION_DRAFT);
        if (compareValues(draft, newValue))
            return;
        selection.remove(SELECTION_DRAFT);
        selection.put(SELECTION_DRAFT, newValue);
        if (notify)
            for (PickerSelectionListeners<SelectionType> listener : selectionListeners)
                listener.onDraftSelectionChangedListener(newValue);
    }

    public void clearDraft() {
        setDraftSelection(null, false);
    }

    public void startDraft() {
        SelectionType current = getCurrentSelection();
        setDraftSelection(current, false);
    }

    public void commitDraft() {
        SelectionType beforeCommit = getCurrentSelection();
        SelectionType draft = getDraftSelection();
        setCurrentSelection(draft, false);
        clearDraft();
        for (PickerSelectionListeners<SelectionType> listener : selectionListeners)
            listener.onDraftCommit(draft, beforeCommit);
    }

    public void revertDraft() {
        clearDraft();
        for (PickerSelectionListeners<SelectionType> listener : selectionListeners)
            listener.onDraftReverted();
    }

    public interface PickerSelectionListeners<SelectionType> {

        void onCurrentSelectionChangedListener(SelectionType newValue, SelectionType oldValue);

        void onDraftSelectionChangedListener(SelectionType newValue);

        void onDraftCommit(SelectionType newValue, SelectionType beforeCommit);

        void onDraftReverted();
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadows/lightbulb/dialogs/base/BaseDialogSelection.kt
package com.github.rooneyandshadows.lightbulb.dialogs.base

import android.os.Bundle
import java.util.ArrayList
import java.util.HashMap

@JvmSuppressWildcards
abstract class BaseDialogSelection<SelectionType>(current: SelectionType?, draft: SelectionType?) {
    private val selectionCurrentKey = "SELECTION_CURRENT"
    private val selectionDraftKey = "SELECTION_DRAFT"
    protected val CURRENT_SELECTION_STATE_KEY = "CURRENT_SELECTION_STATE_KEY"
    protected val DRAFT_SELECTION_STATE_KEY = "DRAFT_SELECTION_STATE_KEY"
    private val selection = HashMap<String, SelectionType?>()
    private val selectionListeners = ArrayList<PickerSelectionListeners<SelectionType>>()
    protected abstract fun compareValues(v1: SelectionType?, v2: SelectionType?): Boolean
    abstract fun hasCurrentSelection(): Boolean
    abstract fun hasDraftSelection(): Boolean
    abstract fun saveState(): Bundle
    abstract fun restoreState(source: Bundle)

    fun getActiveSelection(): SelectionType? {
        return if (hasDraftSelection()) getDraftSelection()
        else getCurrentSelection()
    }

    init {
        selection[selectionCurrentKey] = current
        selection[selectionDraftKey] = draft
    }

    fun getCurrentSelection(): SelectionType? {
        return selection[selectionCurrentKey]
    }

    fun getDraftSelection(): SelectionType? {
        return selection[selectionDraftKey]
    }

    fun addSelectionListeners(selectionListeners: PickerSelectionListeners<SelectionType>) {
        this.selectionListeners.add(selectionListeners)
    }

    fun setCurrentSelection(newValue: SelectionType?) {
        setCurrentSelection(newValue, true)
    }

    fun setDraftSelection(newValue: SelectionType?) {
        setDraftSelection(newValue, true)
    }

    open fun setCurrentSelection(newValue: SelectionType?, notify: Boolean) {
        val current = selection[selectionCurrentKey]
        if (compareValues(current, newValue)) return
        selection.remove(selectionCurrentKey)
        selection[selectionCurrentKey] = newValue
        if (!notify) return
        for (listener in selectionListeners)
            listener.onCurrentSelectionChangedListener(newValue, current)
    }

    open fun setDraftSelection(newValue: SelectionType?, notify: Boolean) {
        val draft = selection[selectionDraftKey]
        if (compareValues(draft, newValue)) return
        selection.remove(selectionDraftKey)
        selection[selectionDraftKey] = newValue
        if (!notify) return
        for (listener in selectionListeners)
            listener.onDraftSelectionChangedListener(newValue, draft)
    }

    fun clearDraft() {
        setDraftSelection(null, false)
    }

    fun startDraft() {
        val current = getCurrentSelection()
        setDraftSelection(current, false)
    }

    fun commitDraft() {
        val beforeCommit = getCurrentSelection()
        val draft = getDraftSelection()
        setCurrentSelection(draft, false)
        clearDraft()
        for (listener in selectionListeners) listener.onDraftCommit(draft, beforeCommit)
    }

    fun revertDraft() {
        clearDraft()
        for (listener in selectionListeners) listener.onDraftReverted()
    }

    interface PickerSelectionListeners<SelectionType> {
        fun onCurrentSelectionChangedListener(newValue: SelectionType?, oldValue: SelectionType?)
        fun onDraftSelectionChangedListener(newValue: SelectionType?, oldValue: SelectionType?)
        fun onDraftCommit(newValue: SelectionType?, beforeCommit: SelectionType?)
        fun onDraftReverted()
    }
}